From 24db99dffd8f232080f4b0b541933a5d5b1812bb Mon Sep 17 00:00:00 2001
From: Junxiao Chang <junxiao.chang@intel.com>
Date: Thu, 27 Feb 2020 16:25:11 +0800
Subject: [PATCH 4/4] drivers - handle workqueue allocation failure case

This change is to fix CVE-2019-16233. If workqueue allocation
fails, driver should handle it in time.

Tracked-On: PKT-3137
Signed-off-by: Junxiao Chang <junxiao.chang@intel.com>
---
 drivers/gpu/drm/radeon/radeon_display.c | 4 ++++
 drivers/scsi/qla2xxx/qla_os.c           | 5 +++++
 2 files changed, 9 insertions(+)

diff --git a/drivers/gpu/drm/radeon/radeon_display.c b/drivers/gpu/drm/radeon/radeon_display.c
index 1455492..eb4ff4f 100644
--- a/drivers/gpu/drm/radeon/radeon_display.c
+++ b/drivers/gpu/drm/radeon/radeon_display.c
@@ -691,6 +691,10 @@ static void radeon_crtc_init(struct drm_device *dev, int index)
 	drm_mode_crtc_set_gamma_size(&radeon_crtc->base, 256);
 	radeon_crtc->crtc_id = index;
 	radeon_crtc->flip_queue = alloc_workqueue("radeon-crtc", WQ_HIGHPRI, 0);
+	if( !radeon_crtc->flip_queue) {
+		kfree(radeon_crtc);
+		return;
+	}
 	rdev->mode_info.crtcs[index] = radeon_crtc;
 
 	if (rdev->family >= CHIP_BONAIRE) {
diff --git a/drivers/scsi/qla2xxx/qla_os.c b/drivers/scsi/qla2xxx/qla_os.c
index e730aab..bb4bccf 100644
--- a/drivers/scsi/qla2xxx/qla_os.c
+++ b/drivers/scsi/qla2xxx/qla_os.c
@@ -461,6 +461,11 @@ static int qla25xx_setup_mode(struct scsi_qla_host *vha)
 			goto fail;
 		}
 		ha->wq = alloc_workqueue("qla2xxx_wq", WQ_MEM_RECLAIM, 1);
+		if(!ha->wq) {
+			ret = -ENOMEM;
+			goto fail;
+		}
+
 		vha->req = ha->req_q_map[req];
 		options |= BIT_1;
 		for (ques = 1; ques < ha->max_rsp_queues; ques++) {
-- 
2.7.4

